hw/dv/dpi/gpiodpi/gpiodpi.c

```diff
@@ -21,83 +21,235 @@
-  assert(rv <= PATH_MAX && rv > 0);
+  int path_len;
+  path_len = snprintf(ctx->dev_to_host_path, PATH_MAX, "%s/%s-read", cwd, name);
+  assert(path_len > 0 && path_len <= PATH_MAX);
+  path_len = snprintf(ctx->host_to_dev_path, PATH_MAX, "%s/%s-write", cwd, name);
+  assert(path_len > 0 && path_len <= PATH_MAX);
 
-  rv = mkfifo(ctx->fifo_pathname, 0644);  // writes are not supported currently
-  if (rv != 0) {
-    fprintf(stderr, "GPIO: Unable to create FIFO at %s: %s\n",
-            ctx->fifo_pathname, strerror(errno));
+  ctx->dev_to_host_fifo = open_fifo(ctx->dev_to_host_path, O_RDWR);
+  if (ctx->dev_to_host_fifo < 0) {
     return NULL;
   }
 
-  ctx->fifo_fd = open(ctx->fifo_pathname, O_RDWR);
-  if (ctx->fifo_fd < 0) {
-    fprintf(stderr, "GPIO: Unable to open FIFO at %s: %s\n", ctx->fifo_pathname,
-            strerror(errno));
+  ctx->host_to_dev_fifo = open_fifo(ctx->host_to_dev_path, O_RDWR);
+  if (ctx->host_to_dev_fifo < 0) {
     return NULL;
   }
 
-  printf(
-      "\n"
-      "GPIO: FIFO pipe created at %s for %d bit wide GPIO. Run\n"
-      "$ cat %s\n"
-      "to observe the output.\n",
-      ctx->fifo_pathname, ctx->n_bits, ctx->fifo_pathname);
+  int flags = fcntl(ctx->host_to_dev_fifo, F_GETFL, 0);
+  fcntl(ctx->host_to_dev_fifo, F_SETFL, flags | O_NONBLOCK);
+
+  print_usage(ctx->dev_to_host_path, ctx->host_to_dev_path, ctx->n_bits);
 
   return (void *)ctx;
 }
 
 void gpiodpi_device_to_host(void *ctx_void, svBitVecVal *gpio_data,
                             svBitVecVal *gpio_oe) {
-  struct gpiodpi_ctx *ctx = (struct gpiodpi_ctx *)ctx_void;
+  gpiodpi_ctx_t *ctx = (gpiodpi_ctx_t *)ctx_void;
   assert(ctx);
 
-  // n_bits > 32 requires more sophisticated handling of svBitVecVal which we
-  // currently don't do (loop through the array of 32 bit values).
-  assert(ctx->n_bits <= 32 && "n_bits must be <= 32");
-
+  // Write 0, 1, or X (when oe is not set) for each GPIO pin, in big endian
+  // order (i.e., the 0th GPIO pin is the last character written).
+  // Finish it with a newline.
   char gpio_str[32 + 1];
-  for (int i = 0; i < ctx->n_bits; i++) {
-    gpio_str[ctx->n_bits - i - 1] = !!(gpio_data[0] & (1 << i)) + '0';
+  char *pin_char = gpio_str;
+  for (int i = ctx->n_bits - 1; i >= 0; --i, ++pin_char) {
+    if (!GET_BIT(gpio_oe[0], i)) {
+      *pin_char = 'X';
+    } else if (GET_BIT(gpio_data[0], i)) {
+      *pin_char = '1';
+    } else {
+      *pin_char = '0';
+    }
   }
+  *pin_char = '\n';
 
-  gpio_str[ctx->n_bits] = '\n';
-  ssize_t written = write(ctx->fifo_fd, gpio_str, ctx->n_bits + 1);
+  ssize_t written = write(ctx->dev_to_host_fifo, gpio_str, ctx->n_bits + 1);
   assert(written == ctx->n_bits + 1);
 }
 
+/**
+ * Parses an unsigned hex number from |text|, advancing it forward as necessary.
+ *
+ * Returns upon encountering any non-hex digit.
+ */
+static uint32_t parse_hex(char **text) {
+  if (text == NULL || *text == NULL) {
+    return 0;
+  }
+
+  uint32_t value = 0;
```

Comment:
Not worth the trouble: there is no UB in overflowing a u32, checking overflow in C is error-prone, and this interface is not exactly what I'd call "untrusted" (since it gets to drive arbitrary GPIO pins...).